static/app/utils/useUndoableReducer.tsx

```diff
@@ -0,0 +1,88 @@
+import {ReducerAction, ReducerState, useReducer} from 'react';
+
+export type UndoableNode<S> = {
+  current: S;
+  next: UndoableNode<S> | undefined;
+  previous: UndoableNode<S> | undefined;
+};
+
+type UndoAction = {
+  type: 'undo';
+};
+
+type RedoAction = {
+  type: 'redo';
+};
+
+type UndoableReducerAction<A> = UndoAction | RedoAction | A;
+
+function isUndoOrRedoAction(
+  action: UndoableReducerAction<any>
+): action is UndoAction | RedoAction {
+  if (action.type) {
+    return action.type === 'undo' || action.type === 'redo';
+  }
+  return false;
+}
+
+function undoableReducer<S, A>(
+  state: UndoableNode<S>,
+  action: UndoableReducerAction<A>
+): UndoableNode<S> {
+  if (!isUndoOrRedoAction(action)) {
+    return state;
+  }
+
+  if (action.type === 'undo') {
+    return state.previous ? state.previous : state;
+  }
+
+  if (action.type === 'redo') {
+    return state.next ? state.next : state;
+  }
+
+  throw new Error('Unreachable case');
```

Comment:
Wouldn't this be cleaner as

```suggestion
  if (action.type === 'undo') {
    return state.previous ? state.previous : state;
  }

  if (action.type === 'redo') {
    return state.next ? state.next : state;
  }

  return state;
```

Reply:
yeah, but you need a proper check for action.type, else TS will complain that type does not exist on type